Documenting the usage of local llms - bring your own Llamafile, Ollama or vLLM model (#806)
mozilla-ai/lumigator: docs/source/user-guides/local-models.md

Lumigator supports running inference and evaluation on any locally hosted models through [Llamafile](https://github.com/Mozilla-Ocho/llamafile), [Ollama](https://ollama.com/search) and [vLLM](https://docs.vllm.ai/en/latest/), thanks to their compatibility with [OpenAI's Completions API Client](https://github.com/openai/openai-python). This guide will walk you through the process of running evaluation on any local model that you bring from these providers (assuming your machine meets the necessary hardware requirements).

Before installation and setup, here are some recommended system requirements:
* Memory (RAM): 8GB minimum, 16GB or more recommended
* Storage: At least 10GB or more of free space
* Processor: A relatively modern CPU with atleast 4 cores

## What You’ll Need
1. A running instance of [Lumigator](../get-started/installation.md).

>     ```{note}

Author: Yes, renders correctly as is. 
> I don't see where the "```" are closing after the console code snippet.

Adding the closing ``` creates a empty artifact somehow!

    **Before to starting up** the Lumigator application, you need to set a value for the `OPENAI_API_KEY` environment variable. This is because all the local model inference tools discussed here are based on OpenAI API compatible client. However, since we are going to run the models locally, this variable can be set to any placeholder value.
    ```console
    user@host:~/lumigator$ export OPENAI_API_KEY="abc123" # pragma: allowlist secret
    ```

> 2. A dataset for experimentation: You can use the {{ '[sample dataset](https://github.com/mozilla-ai/lumigator/blob/{}/lumigator/sample_data/dialogsum_exc.csv)'.format(commit_id) }} provided in the [Lumigator repository](https://github.com/mozilla-ai/lumigator) or upload your own dataset through the [Lumigator UI](../get-started/ui-guide.md#upload-a-dataset).

Author: Yep, it does! And it is already being used in a previous docs page: https://mozilla-ai.github.io/lumigator/get-started/ui-guide.html#upload-a-dataset

1. Run Lumigator Inference. Make the necessary changes to your inference script to point to the local vLLM server and use the correct model:

   ```bash
   #!/bin/bash
   source common_variables.sh

   EVAL_NAME="vLLM HuggingFaceTB/SmolLM2-360M-Instruct"
   EVAL_DESC="Test inference with vLLM's HuggingFaceTB/SmolLM2-360M-Instruct"
   EVAL_MODEL="vllm://HuggingFaceTB/SmolLM2-360M-Instruct" # Format expected vllm://<model_name>, the model_name must be same as one when running the docker container
   EVAL_MODEL_URL="http://localhost:8090/v1" # vLLM setup to run on port 8090

   curl -s "$BACKEND_URL/api/v1/jobs/inference/" \
   -H 'Accept: application/json' \
   -H 'Content-Type: application/json' \
   -d '{
      "name": "'"$EVAL_NAME"'",
      "description": "'"$EVAL_DESC"'",
      "model": "'"$EVAL_MODEL"'",
      "dataset": "'"$EVAL_DATASET_ID"'",
      "max_samples": "'"$EVAL_MAX_SAMPLES"'",
      "model_url": "'"$EVAL_MODEL_URL"'",
      "system_prompt": "'"$EVAL_SYSTEM_PROMPT"'"
   }'
   ```

   Finally execute the bash script:
   ```console
   user@host:~/lumigator$ bash test_local_llm_eval.sh
   ```

> To view the evalution results on the UI, refer to the section [below](#view-evaluation-results)

Author: Removed in favor of the download results script below.

## View Evaluation Results

> To view the evalution results, go to the [Lumigator UI](http://localhost) and navigate to the "Experiments" tab. Then, select the most recent experiment with your local LLM and click "View Results".

Reviewer: I think this section (View evaluation results) will have to wait.
You could direct users to the ray dashboard... or at most help them retrieve the inference results (?).

Author: As discussed, added the download results script instead.

    ```bash
    #!/bin/bash
    BACKEND_URL=http://localhost:8000 # Lumigator runs on port 8000

    # Get the most recently uploaded dataset
    EVAL_DATASET_ID=$(curl -s "$BACKEND_URL/api/v1/datasets/" | grep -o '"id":"[^"]*"' | head -n1 | cut -d'"' -f4)

    # Basic prompt for LLM, summarization task
    EVAL_SYSTEM_PROMPT="You are a helpful assistant, expert in text summarization. For every prompt you receive, provide a summary of its contents in at most two sentences."

    # Run eval on first 10 rows in the csv, set to -1 if you would like to run it for all rows
    EVAL_MAX_SAMPLES="10"
    ```

You have a choice of choosing one among the below-mentioned local LLM tools. We describe the steps to locally stand up your desired model and enable Lumigator to query the local model's inference endpoint.

## Llamafile
[Llamafile](https://github.com/Mozilla-Ocho/llamafile) bundles LLM weights and a specially-compiled version of [llama.cpp](https://github.com/ggerganov/llama.cpp) into a single executable file, allowing users to run large language models locally without any additional setup or dependencies.

### Procedure

1. Download and Setup Llamafile Locally:

>    * Download any model's Llamafile from the official [repo](https://github.com/Mozilla-Ocho/llamafile?tab=readme-ov-file#other-example-llamafiles). For example, `mistral-7b-instruct-v0.2.Q4_0.llamafile` is a 3.85 GB Llamafile to get started with (alternatively `Llama-3.2-1B-Instruct.Q6_K.llamafile` is smaller at 1.11 GB).

Author: Updated text to link to llamafile repo for setup

   * Grant execution permissions: `chmod +x mistral-7b-instruct-v0.2.Q4_0.llamafile`.
   * Start the application locally with `./mistral-7b-instruct-v0.2.Q4_0.llamafile`.

1. Verify Llamafile is Running. You should be able to see it running on
   [localhost:8080](http://localhost:8080/). Note that this is the endpoint that Lumigator will use
   to interact with.

> 1. Run Lumigator Evaluation. Create a new bash script `test_local_llm_eval.sh`:

Reviewer: SDK should be fine, but for UI we can follow up with #858 .

   ```bash
   #!/bin/bash
   source common_variables.sh

   EVAL_NAME="Llamafile mistral-7b-instruct-v0.2"
   EVAL_DESC="Test evaluation with mistral-7b-instruct-v0.2"
   EVAL_MODEL="llamafile://mistral-7b-instruct-v0.2" # Format llamafile://<model_name>
   EVAL_MODEL_URL="http://localhost:8080/v1" # Llamafile runs on port 8080

   # TODO: change to evaluate endpoint after experiments_new migration
   curl -s "$BACKEND_URL/api/v1/jobs/inference/" \
   -H 'Accept: application/json' \
   -H 'Content-Type: application/json' \
   -d '{
      "name": "'"$EVAL_NAME"'",
      "description": "'"$EVAL_DESC"'",
      "model": "'"$EVAL_MODEL"'",
      "dataset": "'"$EVAL_DATASET_ID"'",
      "max_samples": "'"$EVAL_MAX_SAMPLES"'",
      "model_url": "'"$EVAL_MODEL_URL"'",
      "system_prompt": "'"$EVAL_SYSTEM_PROMPT"'"
   }'
   ```

> 1. Finally run the evaluation:

Reviewer: ```suggestion
1. Finally run the inference (i.e. get predictions made by the model you are running locally):
```

   ```console

>    user@host:~/lumigator$ bash test_local_llm_eval.sh

Reviewer: As discussed, this is now tracked in #857.

1. Verify Ollama is Running. An Ollama completions endpoint should start running locally and can be
   verified by visiting [localhost:11434](http://localhost:11434/).

1. Run Lumigator Evaluation. The evaluation steps are similar to earlier but we modify model details
   in the  `test_local_llm_eval.sh` script:

   ```bash
   #!/bin/bash
   source common_variables.sh

   EVAL_NAME="Ollama Llama3.2"
   EVAL_DESC="Test evaluation with Ollama's Llama3.2"
   EVAL_MODEL="ollama://llama3.2" # Format expected ollama://<model_name>, the model_name must be same as one used in the `ollama run <model_name>` command
   EVAL_MODEL_URL="http://localhost:11434/v1" # Ollama runs on port 11434

   # TODO: change to evaluate endpoint after experiments_new migration
   curl -s "$BACKEND_URL/api/v1/jobs/inference/" \
   -H 'Accept: application/json' \
   -H 'Content-Type: application/json' \
   -d '{
      "name": "'"$EVAL_NAME"'",
      "description": "'"$EVAL_DESC"'",
      "model": "'"$EVAL_MODEL"'",
      "dataset": "'"$EVAL_DATASET_ID"'",
      "max_samples": "'"$EVAL_MAX_SAMPLES"'",
      "model_url": "'"$EVAL_MODEL_URL"'",
      "system_prompt": "'"$EVAL_SYSTEM_PROMPT"'"
   }'
   ```

> 1. Run the evaluation:

Reviewer: ```suggestion
1. Run the inference:
```

Author: Replaced evaluation -> inference

   ```bash
   #!/bin/bash
   source common_variables.sh

   EVAL_NAME="vLLM HuggingFaceTB/SmolLM2-360M-Instruct"
   EVAL_DESC="Test evaluation with vLLM's HuggingFaceTB/SmolLM2-360M-Instruct"
   EVAL_MODEL="vllm://HuggingFaceTB/SmolLM2-360M-Instruct" # Format expected vllm://<model_name>, the model_name must be same as one when running the docker container
   EVAL_MODEL_URL="http://localhost:8090/v1" # vLLM setup to run on port 8090

   # TODO: change to evaluate endpoint after experiments_new migration
   curl -s "$BACKEND_URL/api/v1/jobs/inference/" \
   -H 'Accept: application/json' \
   -H 'Content-Type: application/json' \
   -d '{
      "name": "'"$EVAL_NAME"'",
      "description": "'"$EVAL_DESC"'",
      "model": "'"$EVAL_MODEL"'",
      "dataset": "'"$EVAL_DATASET_ID"'",
      "max_samples": "'"$EVAL_MAX_SAMPLES"'",
      "model_url": "'"$EVAL_MODEL_URL"'",
      "system_prompt": "'"$EVAL_SYSTEM_PROMPT"'"
   }'
   ```

1. Run the evaluation:

   ```console
   user@host:~/lumigator$ bash test_local_llm_eval.sh
   ```

To view the evalution results on the UI, refer to the section [below](#view-evaluation-results)

## View Evaluation Results

> To view the evalution results, go to the [Lumigator UI](http://localhost) and navigate to the "Experiments" tab. Then, select the most recent experiment with your local LLM and click "View Results".

Reviewer: ... I'm rather puzzled. Perhaps I've missed a step, but I don't think at this point an experiment will have been created, no? If I missed something, do excuse me and point me to the step 😅

Reviewer: As discussed - you can update this doc to be explicit that it only covers inference and then as soon as the workflow endpoints are public and we get #573 merged, this documentation can also be updated (#857).

# Bring Your Own Local LLMs

Previously we saw how to use Lumigator with models on Hugging Face as well as via APIs hosted by [Open AI and Mistral](../user-guides/inference.md#model-specification). However, sometimes it may be advantageous to run models on-premise as a more cost-effective solution or when dealing with sensitive data or during the early stages of experimentation.

> Lumigator supports running inference and evaluation on any locally hosted models through [Llamafile](https://github.com/Mozilla-Ocho/llamafile), [Ollama](https://ollama.com/search) and [vLLM](https://docs.vllm.ai/en/latest/), thanks to their compatibility with [OpenAI's Completions API Client](https://github.com/openai/openai-python). This guide will walk you through the process of running inference (i.e. get predictions made by the model you are running locally) on any local model that you bring from these providers (assuming your machine meets the necessary hardware requirements).

Reviewer: ```suggestion
Lumigator supports running inference on any locally hosted models through [Llamafile](https://github.com/Mozilla-Ocho/llamafile), [Ollama](https://ollama.com/search) and [vLLM](https://docs.vllm.ai/en/latest/), thanks to their compatibility with [OpenAI's Completions API Client](https://github.com/openai/openai-python). This guide will walk you through the process of running inference (i.e. get predictions made by the model you are running locally) on any local model that you bring from these providers (assuming your machine meets the necessary hardware requirements).
```

Before installation and setup, here are some recommended system requirements:
* Memory (RAM): 8GB minimum, 16GB or more recommended
* Storage: At least 10GB or more of free space

> * Processor: A relatively modern CPU with atleast 4 cores

Reviewer 2: ```suggestion
* Processor: A relatively modern CPU with at least 4 cores
```

## What You’ll Need

1. A running instance of [Lumigator](../get-started/installation.md).

    ```{note}
   Before starting up the Lumigator application, you need to set a value for the `OPENAI_API_KEY` environment variable. This is because all the local model inference tools discussed here are based on OpenAI API compatible client. However, since we are going to run the models locally, this variable can be set to any placeholder value.
      ```console
      user@host:~/lumigator$ export OPENAI_API_KEY="abc123" # pragma: allowlist secret
      ```

> 2. A dataset for experimentation: You can upload the {{ '[sample dataset](https://github.com/mozilla-ai/lumigator/blob/{}/lumigator/sample_data/dialogsum_exc.csv)'.format(commit_id) }} provided in the [Lumigator repository](https://github.com/mozilla-ai/lumigator) or upload your own dataset through the [Lumigator UI](../get-started/ui-guide.md#upload-a-dataset).

Reviewer 2: Minor edit (lowercase Y)

```suggestion
2. A dataset for experimentation: you can upload the {{ '[sample dataset](https://github.com/mozilla-ai/lumigator/blob/{}/lumigator/sample_data/dialogsum_exc.csv)'.format(commit_id) }} provided in the [Lumigator repository](https://github.com/mozilla-ai/lumigator) or upload your own dataset through the [Lumigator UI](../get-started/ui-guide.md#upload-a-dataset).
```

> 3. Create a bash file `common_variables.sh` and initialize the following variables before proceeding with different local LLM tools.

Reviewer 2: ```suggestion
3. Create a bash file `common_variables.sh` and run it to initialize the following variables before proceeding with different local LLM tools.
```

Author: We only create the variables, here I think the "run" happens when in the source `source common_variables.sh` in the inference script below.

   INFERENCE_MODEL_URL="http://localhost:11434/v1" # Ollama runs on port 11434

   curl -s "$BACKEND_URL/api/v1/jobs/inference/" \
   -H 'Accept: application/json' \
   -H 'Content-Type: application/json' \
   -d '{
      "name": "'"$INFERENCE_NAME"'",
      "description": "'"$INFERENCE_DESC"'",
      "model": "'"$INFERENCE_MODEL"'",
      "dataset": "'"$INFERENCE_DATASET_ID"'",
      "max_samples": "'"$INFERENCE_MAX_SAMPLES"'",
      "model_url": "'"$INFERENCE_MODEL_URL"'",
      "system_prompt": "'"$INFERENCE_SYSTEM_PROMPT"'"
   }'
   ```

   Finally execute the bash script:
   ```console
   user@host:~/lumigator$ bash test_local_llm_inference.sh
   ```

Finally, download the results as described [below](#download-inference-results)

## vLLM

[vLLM](https://github.com/vllm-project/vllm) is a high-performance library for LLM inference and serving, featuring optimized memory management techniques. Apart from cloud deployments, it also comes with options to deploy models locally. Moreover, with vLLM you can host any model available on the [HuggingFace Model Hub](https://huggingface.co/models).

> As a pre-requisite, you will need to create an account on HuggingFace and [setup an API token](https://huggingface.co/settings/tokens).

Reviewer: This is a good point and one more reason to link as early as possible to the 3rd party documentation. This doc helps you use lumigator models that you've launched - to launch them, there are 3 possible libraries, each with their documentation, you can use.

> As a pre-requisite, you will need to create an account on HuggingFace and [setup an API token](https://huggingface.co/settings/tokens).

Author: Valid point - there can be a number of different requirements depending on which model you would like to use. I have added a line to refer to vLLM/HF documentation for the specifics
